Add new methods, pause and resume, in I2S driver

kishore-sn opened the pull request:
Add i2s_pause and i2s_resume methods for s5j_i2s driver.
These methods will be used to pause and resume i2s transfer
while trying to recover from xrun error during audio play / record.

seinfra commented:
Code Rule Check OK.

Taejun-Kwon commented:
@GalkinIvan @btheosam could you please check this PR?

sunghan-chang commented:
@kishore-sn To know what you do from commit title clearly, how about below?
```
Add new methods, pause and resume, in I2S driver
```

btheosam commented on os/arch/arm/src/s5j/s5j_i2s.c:
```diff
@@ -1592,6 +1596,63 @@ static uint32_t i2s_txdatawidth(struct i2s_dev_s *dev, int bits)
+	flags = irqsave();
+	s5j_dmastop(priv->txp.dma);
+	modifyreg32(priv->base + S5J_I2S_CON, I2S_CR_FTXURINTEN, 0);
+	modifyreg32(priv->base + S5J_I2S_CON, I2S_CR_TXDMACTIVE, 0);
```
This is not pause. the dma registers should be modified for dma pause. First of all not sure if i2s_pause can be generic method to be added (leave the decision to Ivan).

btheosam commented on os/arch/arm/src/s5j/s5j_i2s.c:
```diff
@@ -1592,6 +1596,63 @@ static uint32_t i2s_txdatawidth(struct i2s_dev_s *dev, int bits)
+	flags = irqsave();
+	s5j_dmastop(priv->rx.dma);
+	modifyreg32(priv->base + S5J_I2S_CON, I2S_CR_FRXOFINTEN, 0);
+	modifyreg32(priv->base + S5J_I2S_CON, I2S_CR_RXDMACTIVE, 0);
```
same as above.

GalkinIvan commented:
This implementation of pause/resume acually disables operatins, not pause.

Taejun-Kwon commented on os/arch/arm/src/s5j/s5j_i2s.c:
```diff
@@ -336,6 +338,8 @@ static const struct i2s_ops_s g_i2sops = {
 	.i2s_send = i2s_send,
 
 	.i2s_stop = i2s_stop,
+	.i2s_pause = i2s_pause,
```
these 2 function should be added in i2s.h too.
I think you missed i2s.h...

kishore-sn commented:
@GalkinIvan  Thanks for pointing out the registers for pause / resume. 
I have modified accordingly. Kindly review.

@sunghan-chang  I modified the commit title as you suggested.

seinfra commented:
Code Rule Check OK.

btheosam commented:
Not all hardware might support dma pause. Please check if this is can be a common solution. What is the alternative if chipset dont support dma pause?

btheosam commented:
HI Kishore, please add your test results as well.

btheosam commented on os/arch/arm/src/s5j/s5j_i2s.c:
```diff
@@ -1592,6 +1596,55 @@ static uint32_t i2s_txdatawidth(struct i2s_dev_s *dev, int bits)
+#ifdef I2S_HAVE_TX_P
+	if (priv->txpenab) {
+		flags = irqsave();
+		modifyreg32(priv->base + S5J_I2S_CON, 0, I2S_CR_TXCHPAUSE);
```
@GalkinIvan If this dma is paused, then data wont flow, would n't that trigger I2S OVF/UND interrupt?

kishore-sn replied:
Since we are pausing the I2S channel itself, there wont be any xrun.

btheosam replied:
@kishore-sn what do you mean  by I2S channel?

btheosam replied:
Okay

btheosam commented on os/arch/arm/src/s5j/s5j_i2s.c:
```diff
@@ -1592,6 +1596,55 @@ static uint32_t i2s_txdatawidth(struct i2s_dev_s *dev, int bits)
+#ifdef I2S_HAVE_RX
+	if (priv->rxenab) {
+		flags = irqsave();
+		modifyreg32(priv->base + S5J_I2S_CON, 0, I2S_CR_RXCHPAUSE);
```
@GalkinIvan If this dma is paused, then data wont flow, would n't that trigger I2S OVF/UND interrupt?

GalkinIvan replied:
If channel operation is paused, no data is transferred, no OVF/UND should happen.

GalkinIvan commented:
Code seems to be fine.
